File: replace_temp_with_query.py
## before

class BookStore(object):
    book_price = 100

    def __init__(self, num):
        self.num = num

    @property
    def price(self):
        """if price * num >= 1000 service price"""
        rate = 1.0
        base_price = self.num * self.__class__.book_price

        if base_price >= 1000:
            rate = 0.9
        return base_price * rate

## after

class BookStore2(object):
    book_price = 100

    def __init__(self, num):
        self.num = num

    @property
    def price(self):
        """if price * num >= 1000 service price"""
        rate = 1.0
        if self._base_price() >= 1000:
            rate = 0.9
        return self._base_price() * rate

    def _base_price(self):
        return self.num * self.__class__.book_price
        
## after2

class BookStore3(object):
    book_price = 100

    def __init__(self, num):
        self.num = num

    @property
    def price(self):
        """if price * num >= 1000 service price"""
        return self._base_price() * self._rate()

    def _rate(self):
        if self._base_price() >= 1000:
            return 0.9
        return 1.0

    def _base_price(self):
        return self.num * self.__class__.book_price


if __name__ == '__main__':
    b = BookStore(10)
    b2 = BookStore2(10)
    b3 = BookStore3(10)

    assert b.price == b2.price == b3.price

File: src/main/c/replace_temp_with_query.cpp
#include <cassert>
#include <vector>
#include <iostream>

using namespace std;

// Before
class BookStore
{
private:
    static const int bookBasePrice = 100;
    int num;

public:
    BookStore(int num);
    int price();
};

BookStore::BookStore(int num)
{
    this->num = num;
}

int BookStore::price()
{
    double rate = 1.0;
    double basePrice = num * bookBasePrice;

    if (basePrice >= 1000) {
        rate = 0.9;
    }

    return basePrice * rate;
}

// After
class BookStore2
{
private:
    static const int bookBasePrice = 100;
    int num;

public:
    BookStore2(int num);
    int price();
    double basePrice();
};

BookStore2::BookStore2(int num)
{
    this->num = num;
}

int BookStore2::price()
{
    double rate = 1.0;

    if (basePrice() >= 1000) {
        rate = 0.9;
    }

    return basePrice() * rate;
}

double BookStore2::basePrice()
{
    return num * bookBasePrice;
}


int
main(void)
{
    BookStore bs(100);
    BookStore2 bs2(100);

    printf("%d\n", bs.price());
    assert(bs.price() == bs2.price());
    return 0;
}
